File: pa1/List.java
//Rahul Patel
//UserID: rpate3
//pa1
//
 class List {
	private class node{
		int data;
		node next;
		node prev;
		
		node(int num){ //constructor
			data = num; 
			next = null;
			prev = null;
		}
		
		//second constructor (dummynode) having links to next and prev
		node(int num, node previous, node post){
			data = num;
			prev = previous;
			next = post; 
		}
		public String toString(){ 
			return "" + data;
			}
	}
	
	//variable declaration
	private static final node OFF = null; 
	private node cursor; 
	private node front; 
	private node back; 
	private int length; 
	
	public List(){ //creates empty list
		cursor = OFF;
		front = null; 
		back = null; 
		length = 0; 
	}
	
	public int length() { //returns number of elements in list
		return length;
	}
	
	//returns index of cursor element
	public int getIndex(){
		int index = -1;  //-1 if it is undefined 
		if(cursor == OFF) return index; 
		for(node temp = front; temp !=null; temp=temp.next){
			index++; 
			if(temp == cursor) break; 
		}
	return index;
	}
	
	//returns front element if length>0
	public int front(){
		if(length() > 0){
			return front.data;
		}
		throw new RuntimeException(); 
	}
	
	//returns back element if length>0 
	public int back() { 
		if(length() > 0){
			return back.data; 
		}
		throw new RuntimeException();
	}
	
	//returns element
	public int getElement(){
		if(length() > 0 && getIndex() >= 0){
			return cursor.data; 
			}
		throw new RuntimeException();
	}
	
	//returns true if this List and L are same integer sequence
	public boolean equals(List L){
		node one = front; 
		node two = L.front; 
		while(one!=null && two!=null){
			if(one.data != two.data) return false; 
			one = one.next; 
			two = two.next; 
		}
		if (one != null || two != null) return false; 
		
		return true; 
	}
	
	//Re-sets this List to the empty state.
	public void clear(){
		front = null; 
		back = null; 
		cursor = OFF; 
		length = 0; 
	}
	
	//moves the cursor to the element 
	// at index i
	public void moveTo(int i){
		if (i>=0 && i<= length()-1){
			cursor = front; 
			for(; i>0; i--) cursor = cursor.next; 
		}else {
			cursor = OFF;
		}
	}
	
	//This operation is
	// equivalent to moveTo(getIndex()-1)
	public void movePrev(){
		int index = getIndex();
		if(index >=0 && index<= length()-1){
			cursor = cursor.prev; 
		} else{
			cursor = OFF;
		}
	}

	//moves cursor one step forward
	public void moveNext(){
		int index = getIndex(); 
		if(index >=0 && index < length()-1){
			cursor = cursor.next; 
		} else{
			cursor = OFF; 
		}
	}
	
	//prepend list, inserts new element before front element in this List
	public void prepend(int data){
		length++; 
		node temp = new node(data,null,front); 
		if(front != null){
			front.prev = temp; 
		} else{
			back = temp;
		}
		front = temp;
	}
	
	//appends list, inserts new element after last element in this List
	public void append(int data){
		length ++; 
		node temp = new node(data, back, null); 
		if(back != null){
			back.next= temp; 
		} else{
			front = temp;
		}
		back=temp; 
	}
	
	
	
	
	// Inserts new element before cursor element in the list
	public void insertBefore(int data){
		if(!(length()>0 && getIndex() >=0)) return;
		length ++; 
		node temp = new node(data, cursor.prev, cursor);
		if(cursor.prev != null){
			cursor.prev.next = temp; 
		} else {
			front = temp; 
		}
		cursor.prev = temp; 
	}
	
	// Inserts new element after cursor element in this List
	public void insertAfter (int data){ 
		if(!(length() > 0 && getIndex() >= 0)) return; 
		length ++; 
		node temp = new node(data, cursor, cursor.next);
		if (cursor.next != null){
			cursor.next.prev = temp; 
		} else{
			back = temp; 
		}
		cursor.next = temp; 
	}
	
	//deletes front node
	public void deleteFront(){
		if (length() > 0){
			length--;
			if(cursor == front) cursor = OFF; 
			front = front.next; 
			front.prev = null; 
		}
	}
	
	//deletes backnode
	public void deleteBack(){
		if(length() > 0){
			length--;
			if(cursor == back) cursor = OFF; 
			back = back.prev; 
			back.next = null;
		}
	}
	
	 //Deletes cursor element in this List.
	public void delete(){
		if(length() > 0 && getIndex() >= 0){
			length--; 
			cursor.next.prev = cursor.prev; 
			cursor.prev.next = cursor.next; 
			cursor = OFF; 
		}
		
	}
	
	 //Creates a string 
	 // consisting of a space separated sequence of integers
	 // with front on the left and back on the right
	public String toString(){
		String result =""; 
		node temp = front; 
		while(temp != null){
			result += temp +" "; 
			temp = temp.next; 
		}
		return result; 
	}
	
	// Returns a new list representing the same integer sequence as this
	// list.
	public List copy(){ 
		List copy = new List(); 
		for (node temp = front; temp != null; temp= temp.next){ 
			copy.append(temp.data); 
		}
		return copy;
		}
	
	// Returns a new List which is the concatenation of 
	// this list followed by L. 
	List concat(List L){
		List linked = copy(); 
		for(node temp = L.front; temp != null; temp= temp.next){
			linked.append(temp.data);
		}
		return linked;
	}
 		
 }
 

File: pa2/List.c
//Rahul Patel
//rpate3
//pa2

#include <stdlib.h>
#include "List.h"

//node points to nodeObj
typedef struct nodeObj* node; 

struct nodeObj{
	node next; 
	node prev; 
	int data;
};

//ListObj type
typedef struct ListObj{
	node cursor; 
	node front;
	node back; 
	int length; 
	int index;
}ListObj; 

//Creates a new empty list 
List newList(){
	List listA= malloc (sizeof(struct ListObj)); 
	listA->cursor = NULL;
	listA->back = NULL;
	listA->front = NULL; 
	listA->length = 0; 
	listA->index = -1; 
	return listA;
}
//frees memory locations
void freeList(List *pL){
	if((*pL) == NULL) return; 
	clear(*pL); 
	free(*pL);
	*pL = NULL; 
}

//Returns number of elements in this list
int length(List L){ 
	if(L==NULL) return -1; 
	return L->length; 
}
//Returns the index of the cursor element in this list, or returns -1 if the cursor 
//element is undefined
int getIndex(List L){
	if (L == NULL) return -1;
	return L->index;
}

//Returns front element in this list
int front(List L){
	if(L == NULL) return -1;
	if(L->length >0){
		return L->front->data;
	}
	return -1;
}

//Returns back element in this list
int back(List L){ 
	if(L == NULL) return -1; 
	if(L->length>0){
		return L->back->data; 
	}
	return -1;
}

//Returns cursor element in this list
int getElement (List L){ 
	if(L == NULL) return -1; 
	if(L->length > 0 && L->index >=0) { 
		return L->cursor->data; 

	}
	return -1;
}

//Returns true of this List and L are the same integer sequence. 
//The cursor is ignored in both lists
int equals(List A, List B){
	if(A == NULL || B == NULL) return -1; 
	
	node one =  A-> front;
	node two =  B-> front; 
	while(one!=NULL && two != NULL){
		if(one->data != two->data) return 0; 
		one = one->next; 
		two = two->next; 
	}
	if(one != NULL || two != NULL) return 0; 
	return 1; 
}

//Re-sets this List to the empty state
void clear(List L){
	if(L==NULL) return; 
	node front = L->front; 
	while(front != NULL){
		node temp = front->next; 
		free(front);
		front = temp; 
	}  
	L->front = NULL; 
	L->back = NULL; 
	L->cursor = NULL; 
	L->length = 0;
	L->index = -1;
}

//Moves the cursor to the element at index i, otherwise the cursor becomes undefined
void moveTo(List L, int i){
	if(L==NULL) return; 
	
	L->cursor = L->front; 
	L->index = i;
	if(i>=0 && i<L->length){
		for(; i>0; i--){
			L->cursor= L->cursor->next; 
		}
	}else{
		L->cursor = NULL; 
		L->index =-1;
	}
}

//Moves the cursor one step toward the front of the list 
void movePrev(List L){ 
	if(L==NULL) return;
	if(L->index>=0 && L->index< L->length){ 
		L->cursor = L->cursor->prev;
		L->index--; 
	}else{ 
		L->cursor = NULL;
		L->index = -1;
	}
}

//Moves the cursor one step toward the back of the list
void moveNext(List L){ 
	if(L==NULL) return; 
	if(L->index>=0 && L->index< L->length-1){ 
		L->cursor = L->cursor->next;
		L->index++;  
	}else{ 
		L->cursor = NULL;
		L->index = -1;
	}
}

//Inserts a new element before front element in this List
void prepend(List L, int data){
	L->length++; 
	node node= malloc(sizeof(struct nodeObj)); 
	node ->data = data; 
	node ->prev = NULL; 
	node ->next = L->front; 
	
	if(L->front == NULL){ 
		L->back = node; 
	}else{ 
		L->front->prev = node; 
	} 
	L->front = node;
	if(L->index >=0) L->index++; 
}

//Inserts new element after back element in this List
void append(List L, int data){ 
	if(L==NULL) return; 
	L->length++; 
	node node= malloc(sizeof(struct nodeObj)); 
	node-> data = data;
	node-> prev = L->back;
	node->next = NULL; 

	if(L->back == NULL){
		L->front = node; 
	}else {
		L->back->next = node;		
	}
	L->back = node;   
}

//Inserts new element before cursor element in this List
void insertBefore(List L, int data){
	if(L==NULL) return;
	if(!(L->length>0 && L->index>=0)) return; 
	L->length++;
	
	node node= malloc(sizeof(struct nodeObj)); 
	node->data = data; 
	node->prev = L->cursor->prev; 
	node->next = L->cursor;
	if(node->prev != NULL){
		L->cursor->prev->next = node; 
	}else{
		L->front = node; 
	}
	L->cursor->prev = node;
	L->index++; 
}

//Inserts new element after cursor element in this List 
void insertAfter(List L, int data){
	if(L==NULL) return; 
	if(!(L->length>0 && L->index>=0)) return; 
	L->length++; 
	
	node node= malloc(sizeof(struct nodeObj));
	node->data = data; 
	node->prev = L->cursor; 
	node->next = L->cursor->next; 

	if(node->next != NULL){
		L->cursor->next->prev = node; 
	}else{
		L->back = node; 
	}
	L->cursor->next = node; 
}

//Deletes the front element in this List
void deleteFront(List L){
	if(L->length>0){
		L->length--;
		if(L->cursor == NULL) return; 
		if(L->back == L->front) L->back = NULL; 
		node temp = L->front;
		L->front = L->front->next;
		free(temp); 
		L->front->prev = NULL;
		if(L->index != -1) L->index--; 
	}
}

//Deletes the back element in this List
void deleteBack(List L){
	if(L->length>0){
		L->length--; 
		if(L->index == L->length-1){
			 L->cursor == NULL; 
			 L->index = -1; 
		}
		if(L->back == L->front) L->front = NULL; 
		node temp = L->back;
		L->back = L->back->prev;
		free(temp);  
		L->back->next = NULL; 	
	}
}

//Deletes cursor element in this List
void delete(List L){
	if(L->length > 0 && L->index > 0){
		L->length--; 
		L->cursor->next->prev = L->cursor->prev;
		L->cursor->prev->next = L->cursor->next; 
		L->cursor = NULL; 
		L->index = -1; 
	}
}

//Prints the L to the file pointed to by out
void printList(FILE* out, List L){
	node temp = L->front; 
	while(temp!=NULL){
		printf("%d ",temp->data);
		temp = temp->next; 
	}
}

//copies List
List copyList(List L){
	List copy = newList(); 
	for(node temp = L->front; temp != NULL; temp = temp->next){
		append(copy, temp->data); 
	}
	return copy; 
}
